src/nccl_ofi_api.c

```diff
@@ -308,57 +315,139 @@ ncclResult_t nccl_net_ofi_regMr(void *comm, void *data, size_t size, int type,
+				if (!ptr) {
+					NCCL_OFI_WARN("Unable to grow cache");
+					goto out;
+				}
+				cache->slots = ptr;
+			}
+			assert(cache->slots);
+			memmove(cache->slots+slot+1, cache->slots+slot, (cache->size - slot) * sizeof(nccl_ofi_reg_entry_t*));
+			cache->slots[slot] = calloc(1, sizeof(nccl_ofi_reg_entry_t));
+
+			entry = cache->slots[slot];
+			/* Register with network */
+			switch (base_comm->type) {
+			case NCCL_NET_OFI_SEND_COMM:;
+				nccl_net_ofi_send_comm_t *send_comm =
+					(nccl_net_ofi_send_comm_t *)base_comm;
+				ret = send_comm->regMr(send_comm, data, size, type, &entry->handle);
+				break;
+			case NCCL_NET_OFI_RECV_COMM:;
+				nccl_net_ofi_recv_comm_t *recv_comm =
+					(nccl_net_ofi_recv_comm_t *)base_comm;
+				ret = recv_comm->regMr(recv_comm, data, size, type, &entry->handle);
+				break;
+			default:
+				NCCL_OFI_WARN("Unexpected communicator type. Communicator type: %d",
+					      base_comm->type);
+				ret = -EINVAL;
+				goto out;
+			}
+
+			entry->addr = addr;
+			entry->pages = pages;
+			entry->refcnt = 1;
+			cache->used++;
+			*mhandle = entry->handle;
+			goto out;
+		} else if ((addr >=  cache->slots[slot]->addr) &&
+			   ((addr -  cache->slots[slot]->addr)/system_page_size + pages) <=  cache->slots[slot]->pages) {
+			/* Cache hit */
+			NCCL_OFI_TRACE(NCCL_NET, "Found MR handle for %p in cache slot %d", data, slot);
+			cache->slots[slot]->refcnt++;
+			*mhandle = cache->slots[slot]->handle;
+			goto out;
+		}
 	}
 
+out:
+	nccl_net_ofi_mutex_unlock(&cache->lock);
 	return nccl_net_ofi_retval_translate(ret);
 }
 
 ncclResult_t nccl_net_ofi_deregMr(void *comm, void *mhandle)
 {
+	nccl_net_ofi_device_t *dev;
+	nccl_ofi_mr_cache_t *cache;
+	int ret = 0;
+
 	/* Retrieve and validate comm */
 	nccl_net_ofi_comm_t *base_comm =
 		(nccl_net_ofi_comm_t *)comm;
 	if (OFI_UNLIKELY(base_comm == NULL)) {
 		NCCL_OFI_WARN("Invalid comm object provided");
 		return ncclInternalError;
 	}
-
-	int ret = 0;
-
-	switch (base_comm->type) {
-	case NCCL_NET_OFI_SEND_COMM:;
-		nccl_net_ofi_send_comm_t *send_comm =
-			(nccl_net_ofi_send_comm_t *)base_comm;
-		ret = send_comm->deregMr(send_comm, mhandle);
-		break;
-	case NCCL_NET_OFI_RECV_COMM:;
-		nccl_net_ofi_recv_comm_t *recv_comm =
-			(nccl_net_ofi_recv_comm_t *)base_comm;
-		ret = recv_comm->deregMr(recv_comm, mhandle);
-		break;
-	default:
-		NCCL_OFI_WARN("Unexpected communicator type. Communicator type: %d",
-			      base_comm->type);
-		ret = -EINVAL;
-		break;
+	dev = base_comm->ep->device;
+	cache = &dev->mr_cache;
+
+	for (int i = 0; i < cache->used; i++) {
+		if (mhandle == cache->slots[i]->handle) {
+			/* Keep the entry alive for other users */
+			if (--cache->slots[i]->refcnt)
+				goto out;
+
+			switch (base_comm->type) {
+			case NCCL_NET_OFI_SEND_COMM:;
+				nccl_net_ofi_send_comm_t *send_comm =
+					(nccl_net_ofi_send_comm_t *)base_comm;
+				ret = send_comm->deregMr(send_comm, mhandle);
+				goto out;
```

Comment:
Not seeing why `goto out` is the right thing to do here. Don't we want to continue to free the slot and defrag cache?